util/rate_limiter.cc

```diff
@@ -245,7 +307,10 @@ void GenericRateLimiter::Refill() {
     while (!queue->empty()) {
       auto* next_req = queue->front();
       if (available_bytes_ < next_req->request_bytes) {
-        // avoid starvation
+        // Grant partial request_bytes to avoid starvation of requests
+        // that become asking for more bytes than available_bytes_
+        // due to dynamically reduced rate limiter's bytes_per_second that
+        // leads to reduced refill_bytes_per_period hence available_bytes_
         next_req->request_bytes -= available_bytes_;
```

Comment:
It feels kind of odd that we drain all available_bytes_ in Refill() but not when inserting an initial item onto the queue. It seems changing that would make it fairer in terms of giving bytes to earlier requests first. I think it would also allow us to set `available_bytes_ = refill_bytes_per_period` at the start of `Refill()` rather than the quite confusing `available_bytes_ += refill_bytes_per_period`.

Reply:
> It seems changing that would make it fairer in terms of giving bytes to earlier requests first.

It would also allow us to answer this question positively: https://github.com/facebook/rocksdb/blob/ab7f7c9e497a18c48837556bf2e68b4cc5d2a498/util/rate_limiter.cc#L128-L129. I think right now a big request (bytes > available_bytes) can be enqueued, then a small request (bytes <= available_bytes) can be permitted even though somebody's waiting.

Reply:
This is a suggestion for a follow-up.

Reply:
Done: marked it as part of my learning task